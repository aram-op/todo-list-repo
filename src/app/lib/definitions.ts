export type Todo =  {
    id: string,
    title: string,
    is_completed: boolean
}