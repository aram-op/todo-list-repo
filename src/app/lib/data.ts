import {createPool} from '@vercel/postgres';
import {Todo} from '@/app/lib/definitions';

const pool = createPool({
    connectionString: process.env.NEXT_PUBLIC_POSTGRES_URL,
});

export async function fetchFilteredTodos(query: string) {

    try {
        const todos = await pool.sql<Todo>`
            SELECT * FROM todos WHERE UPPER(title) LIKE UPPER(${`%${query}%`});
        `;

        return todos.rows;
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todos.');
    }
}

export async function fetchTodoById(id: string) {
    try {
        const result = await pool.sql<Todo>`
            SELECT * FROM todos WHERE id = ${id};
        `;

        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todo.');
    }
}

export async function updateTodo(todo: Todo) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${todo.title}`},
             is_completed = ${`${todo.is_completed}`}
             WHERE id = ${`${todo.id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function updateTodoTitle(id: string, title: string) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${title}`},
             WHERE id = ${`${id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function deleteTodoById(id: string) {
    try {
        const result = await pool.sql`
              DELETE FROM todos WHERE id = ${`${id}`};
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to delete todo.');
    }
}

export async function addTodo(title: string) {
    try {
        const result = await pool.sql`
            INSERT INTO todos (title, is_completed)
             VALUES
             (${title}, false)
             ON CONFLICT (id) DO NOTHING;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to add todo.')
    }
}