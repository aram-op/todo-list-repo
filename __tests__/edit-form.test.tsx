import {act, fireEvent, render, screen, waitFor} from '@testing-library/react';
import {updateTodo} from '@/app/lib/data';
import {redirect} from 'next/navigation';
import EditForm from '@/app/ui/edit form/edit-form';

jest.mock('@/app/lib/data', () => ({
    updateTodo: jest.fn(),
}));

jest.mock('next/navigation', () => ({
    redirect: jest.fn(),
}));

describe('EditForm component', () => {
    const mockTodo = {
        id: '1',
        title: 'Buy a cake',
        is_completed: false,
    };

    it('should render the form', () => {
        render(<EditForm todo={mockTodo}/>);

        const input = screen.getByDisplayValue(mockTodo.title);
        expect(input).toBeInTheDocument();

        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
    });

    it('should handle form submission', async () => {
        render(<EditForm todo={mockTodo}/>);

        const input = screen.getByDisplayValue(mockTodo.title);
        const button = screen.getByRole('button');

        fireEvent.change(input, {target: {value: 'A new mock title'}});

        fireEvent.click(button);

        await waitFor(() => {
            expect(updateTodo).toHaveBeenCalledWith({...mockTodo, title: 'A new mock title'});

            expect(redirect).toHaveBeenCalledWith('/');
        });
    });

    it('should display error message if title is empty', async () => {
        render(<EditForm todo={mockTodo}/>);

        const input = screen.getByDisplayValue(mockTodo.title);
        const button = screen.getByRole('button');

        act(() => {
            fireEvent.change(input, {target: {value: ''}});

            fireEvent.click(button);
        });

        await waitFor(() => {
            expect(screen.getByText('Title field cannot be empty!')).toBeInTheDocument();
        });
    });

    it('should clear the error if the field is not empty', async () => {
        render(<EditForm todo={mockTodo}/>);

        const input = screen.getByDisplayValue(mockTodo.title);

        act(() => {
            fireEvent.change(input, {target: {value: ''}});
            fireEvent.click(screen.getByRole('button'));
        });

        await waitFor(() => {
            expect(screen.getByText('Title field cannot be empty!')).toBeInTheDocument();

            fireEvent.change(input, {target: {value: 'A new mock todo title'}});

            expect(screen.queryByText('Title field cannot be empty!')).not.toBeInTheDocument();
        });
    });

    it('should match snapshot', () => {
        const {asFragment} = render(<EditForm todo={mockTodo}/>);

        expect(asFragment).toMatchSnapshot();
    });
});
