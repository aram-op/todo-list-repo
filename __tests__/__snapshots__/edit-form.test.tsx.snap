// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm component should match snapshot 1`] = `[Function]`;
